### PR TITLE
Low-level functions for modifying attributes hash/linked list in WDL++ object store

### PR DESCRIPTION
**Pull request by ela-ine**
The new `attributes` module will allow access and modification to attributes of `object_t`. This is to assist in parsing objects in the new WDL++ format, addressing #687. New files include documentation, implementation, and testing.

The attribute struct, first introduced in object.h from #707, will be a hash table and will have the option to be a doubly linked list. The struct takes in an id and a piece of data. I've modified the `data` element to be a `void *` rather than an `enum` to allow more flexibility for other teams. 

I've written the following functions to modify the hash and linked list:
- add to attributes hash
- find in attributes hash
- free & delete attribute from hash and linked list
- append attribute to linked list
- free linked list
- free entire hash & associated lists

As well as some very basic interface functions that access elements in an attribute struct.

### CODE REVIEW

File: include/wdl/attributes.h [hunk 1]
```diff
@@ -0,0 +1,123 @@
+#include "common/common.h"
+#include "common/uthash.h"
+#include "common/utlist.h"
+
+#ifndef INCLUDE_ATTRIBUTES_H
+#define INCLUDE_ATTRIBUTES_H
+#define MAXLEN_ID 60 // ID strings for objects
+
+typedef struct obj obj_t; // forward declaration so attribute_t can use
```
Comment (noahklowden) [inline include/wdl/attributes.h:9]: This needs to be changed to object_t.

**Comment by ela-ine** [comment]
Potential backlog issue: if head of linked list is removed, next element should be added to hash?

File: include/wdl/attributes.h [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+#define MAXLEN_ID 60 // ID strings for objects
+
+typedef struct obj object_t; // forward declaration so attribute_t can use
+
+/*
+ * attribute_t: the attributes stored within an object
+ */
+typedef struct attr
+{
+    //the attribute's id.
```
Comment (MaxineK36) [inline include/wdl/attributes.h:16]: These should be block comments

File: include/wdl/object.h [hunk 3]
```diff
@@ -5,13 +5,12 @@
 #include <stdio.h>
 #include "common/common.h"
 #include "common/uthash.h"
-#include "common/utlist.h"
+//#include "common/utlist.h"
```
Comment (MaxineK36) [inline include/wdl/object.h:8]: why is this commented out? it should be removed entirely or uncommented

File: src/wdl/src/attributes.c [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+/* See attributes.h for documentation */
+obj_attr_t *add_attribute(obj_attr_t **attrs, char *id, void * d)
+{
+    if (id == NULL) {
+        return NULL;
+    }
+    obj_attr_t *new = find_attr(attrs, id); // see if key already exists in hash
+    if (new == NULL) {
+        new =  new_attr(id, d);
+        if (new == NULL) return NULL;
```
Comment (MaxineK36) [inline src/wdl/src/attributes.c:44]: please make this if statement two lines

File: src/wdl/src/attributes.c [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    }
+    DL_DELETE(head, a);
+    free(a);
+    return SUCCESS;
+}
+
+/* See attributes.h for documentation */
+int free_attr_list(obj_attr_t **attrs, obj_attr_t *head)
+{
+    if ((attrs == NULL)|(head == NULL)) return FAILURE;
```
Comment (MaxineK36) [inline src/wdl/src/attributes.c:106]: This should be a `||` not a `|`

File: src/wdl/src/attributes.c [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    
+    HASH_DEL(*attrs, head);
+    
+    obj_attr_t *elt, *tmp;
+    DL_FOREACH_SAFE(head, elt, tmp) {
+        if (head == NULL) return SUCCESS;
```
Comment (MaxineK36) [inline src/wdl/src/attributes.c:112]: this if statement should be two lines

File: src/wdl/src/objstore.c [hunk 7]
```diff
@@ -50,12 +50,13 @@ int add_objstore(objstore_t **obj_store, object_t *o)
 int free_objstore(objstore_t **obj_store, objstore_t *store)
 {
     HASH_DEL(*obj_store, store);
+    free(store->o);
```
Comment (MaxineK36) [inline src/wdl/src/objstore.c:53]: should this be calling some object free function?

File: tests/wdl/test_attributes.c [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+/* ---------- HASH FUNCTIONS ---------- */
+
+Test(attributes, new_failure)
+{
+    int data = 100;
+    obj_attr_t *res = new_attr(NULL, &data);
+    cr_assert_null(res, "new_attr() failed - returned value when NULL expected");
+}
+
+Test(attributes, new_success)
```
Comment (MaxineK36) [inline tests/wdl/test_attributes.c:15]: These should be three tests if they're testing 3 different things

File: src/wdl/src/objstore.c (hunk 7 above)
Comment (ela-ine) [reply src/wdl/src/objstore.c:53]: I don't think object_free has been implemented as of yet?

Comment (MaxineK36) [reply src/wdl/src/objstore.c:53]: Ok; in that case, can you add a comment making a note of that TODO?

Comment (MaxineK36) [reply src/wdl/src/objstore.c:53]: @ela-ine

**Comment by ela-ine** [comment]
Addressing #687, the new attributes module will assist in parsing the WDL++ format and allow other teams to access the attributes associated with each object. As a hash and a linked list, the new attribute struct grants increased flexibility to both current features and future features.

**Comment by MaxineK36** [comment]
Issue Score: ✔️++

Comments:
Great work! The bulleted list of what was accomplished is very helpful, as are the links to relevant issues.

**Pull request merged**